Bluetooth: Host: Add support for Advertising Coding Selection

Thomas-Deppe opened the pull request:
Adds API for Advertising Coding Selection.

Introduces two new advertising options to configure the advertiser's requirement concerning coding scheme when LE Coded PHY is configured. While the Bluetooth v6.0 specification makes a distinction betweeen preferred and required advertising PHY options, a simplification is made to only expose the required PHY options. Inline with how LE Coded PHY is implemented; this API will set both the primary and secondary advertising PHY's to the same coding scheme.

The support is enabled by CONFIG_BT_EXT_ADV_CODING_SELECTION, and requires a controller that selects CONFIG_BT_CTLR_ADV_EXT_CODING_SELECTION_SUPPORT.

jhedberg commented on subsys/bluetooth/Kconfig.adv:
```diff
@@ -68,6 +68,14 @@ config BT_PER_ADV_SYNC_RSP
 
+config BT_EXT_ADV_CODING_SELECTION
+	bool "Advertising Coding Selection support"
+	depends on BT_CTLR_PHY_CODED
```
This doesn't seem right. You could have a controller that doesn't set `BT_CTLR` or any related options, yet it supports this feature (e.g. a controller behind some physical HCI transport). Most likely thisould be:
```
depends on !HAS_BT_CTLR || BT_CTLR_PHY_CODED
```

That's the pattern that's used elsewhere too in host side Kconfig definitions.

Thomas-Deppe replied:
Thank you! Sorry about that! I have made the change now :)

cvinayak commented on tests/bsim/bluetooth/host/adv/extended/prj_advertiser.conf:
```diff
@@ -3,3 +3,5 @@ CONFIG_BT_EXT_ADV=y
 CONFIG_BT_MAX_CONN=1
 CONFIG_BT_PERIPHERAL=y
 CONFIG_BT_DEVICE_NAME="test_ext_adv"
+CONFIG_BT_CTLR_PHY_CODED=y
+CONFIG_BT_EXT_ADV_CODING_SELECTION=y
```
Add conf in `tests/bluetooth/init`  build-only to cover the new feature (with/without).

Thomas-Deppe replied:
I think I have managed to implement this as you suggested, but please let me know. Thank you again for the help :)

cvinayak commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1171,6 +1191,22 @@ static int le_ext_adv_param_set(struct bt_le_ext_adv *adv,
 		cp->prim_adv_phy = BT_HCI_LE_PHY_CODED;
 		cp->sec_adv_phy = BT_HCI_LE_PHY_CODED;
+
+#if defined(CONFIG_BT_EXT_ADV_CODING_SELECTION)
+		if (opcode == BT_HCI_OP_LE_SET_EXT_ADV_PARAM_V2) {
+			/* Default to no required coding scheme */
+			uint8_t adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_NO_REQUIRED;
+
+			if (param->options & BT_LE_ADV_OPT_REQUIRE_S8_CODING) {
+				adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_REQUIRE_S8;
+			} else if (param->options & BT_LE_ADV_OPT_REQUIRE_S2_CODING) {
+				adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_REQUIRE_S2;
+			}
+
+			cp->prim_adv_phy_opt = adv_phy_opt;
+			cp->sec_adv_phy_opt = adv_phy_opt;
+		}
+#endif
```
any reason not to use `IS_ENABLED` construct here?

cvinayak commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1171,6 +1191,22 @@ static int le_ext_adv_param_set(struct bt_le_ext_adv *adv,
+#if defined(CONFIG_BT_EXT_ADV_CODING_SELECTION)
+		if (opcode == BT_HCI_OP_LE_SET_EXT_ADV_PARAM_V2) {
+			/* Default to no required coding scheme */
+			uint8_t adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_NO_REQUIRED;
+
+			if (param->options & BT_LE_ADV_OPT_REQUIRE_S8_CODING) {
+				adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_REQUIRE_S8;
+			} else if (param->options & BT_LE_ADV_OPT_REQUIRE_S2_CODING) {
+				adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_REQUIRE_S2;
+			}
```
```suggestion
			uint8_t adv_phy_opt;

			if (param->options & BT_LE_ADV_OPT_REQUIRE_S8_CODING) {
				adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_REQUIRE_S8;
			} else if (param->options & BT_LE_ADV_OPT_REQUIRE_S2_CODING) {
				adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_REQUIRE_S2;
			} else {
				adv_phy_opt = BT_HCI_LE_ADV_PHY_OPTION_NO_REQUIRED;
			}
```

cvinayak replied:
`_PREFERS_` can be a not supported, but do not see passing them over HCI is not too much of work in this PR.

cvinayak commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1121,20 +1121,40 @@ static int le_ext_adv_param_set(struct bt_le_ext_adv *adv,
 	uint16_t props = 0;
 
-	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_EXT_ADV_PARAM, sizeof(*cp));
+	if (IS_ENABLED(CONFIG_BT_EXT_ADV_CODING_SELECTION) &&
+		BT_FEAT_LE_ADV_CODING_SEL(bt_dev.le.features)) {
```
```suggestion
	if (IS_ENABLED(CONFIG_BT_EXT_ADV_CODING_SELECTION) &&
	    BT_FEAT_LE_ADV_CODING_SEL(bt_dev.le.features)) {
```

cvinayak commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -784,6 +784,40 @@ enum {
 	 * @note Mutually exclusive with BT_LE_ADV_OPT_USE_IDENTITY.
 	 */
 	BT_LE_ADV_OPT_USE_NRPA = BIT(19),
+
+	/**
+	 * @brief Configures the advertiser to use the S=2 coding scheme for
+	 * LE Coded PHY.
+	 *
+	 * Sets the advertiser's required coding scheme to S=2, which is one
+	 * of the coding options available for LE Coded PHY. The S=2 coding
+	 * scheme offers higher data rates compared to S=8, with a trade-off
+	 * of reduced range. The coding scheme will only be set if both the
+	 * primary and secondary advertising channels indicate LE Coded Phy.
+	 * Additionally, the Controller must support the LE Feature Advertising
+	 * Coding Selection. If these conditions are not met, it will default to
+	 * no required coding scheme.
+	 *
+	 * @note Requires @kconfig{BT_EXT_ADV_CODING_SELECTION}
+	 */
+	BT_LE_ADV_OPT_REQUIRE_S2_CODING = BIT(20),
+
+	/**
+	 * @brief Configures the advertiser to use the S=8 coding scheme for
+	 * LE Coded PHY.
+	 *
+	 * Sets the advertiser's required coding scheme to S=8, which is one
+	 * of the coding options available for LE Coded PHY. The S=8 coding
+	 * scheme offers increased range compared to S=2, with a trade-off
+	 * of lower data rates. The coding scheme will only be set if both the
+	 * primary and secondary advertising channels indicate LE Coded Phy.
+	 * Additionally, the Controller must support the LE Feature Advertising
+	 * Coding Selection. If these conditions are not met, it will default to
+	 * no required coding scheme.
+	 *
+	 * @note Requires @kconfig{BT_EXT_ADV_CODING_SELECTION}
+	 */
+	BT_LE_ADV_OPT_REQUIRE_S8_CODING = BIT(21),
```
Adding the Host `PREFERS`  bit mask too (do not see it is too much to assign and provide it to the Controller) will ensure future enhancements to support do not have to define the enum away from this new BIT 20/21 values (or have breaking API changes).

Thomas-Deppe replied:
Hi! Thanks for the feedback! I can do that, I guess then bits 20-23 will be taken up by the options. It is not much work to add the support, I was just trying to reduce the options fields :)

rugeGerritsen replied:
Do we foresee that we need to have options for both "prefer" and "require"? I believe that adds additional complexity and controller implementation dependencies. By just adding "require" we keep it simple. 

For the same reason both primary and secondary PHY are set to Coded when Coded is being used

cvinayak replied:
> Do we foresee that we need to have options for both "prefer" and "require"? 

As the HCI_LE_Set_Default_PHY command sets the preferred PHY, "require" PHY makes sense only for non-connectable advertising.

> For the same reason both primary and secondary PHY are set to Coded when Coded is being used
That is ok for now, as this PR is not changing `struct bt_le_adv_param` definition, right?

rugeGerritsen replied:
>"require" PHY makes sense only for non-connectable advertising. 

Could you elaborate on that? This API is to set the coding scheme for advertising only. It doesn't influence phys for the connections

Thomas-Deppe replied:
I will implemented the PREFERS bit mask, depending on the outcome of this discussion :)

cvinayak replied:
What I mean is, as a peripheral device that advertises for a connection with S=2 required, on being connected can only use `struct bt_conn_le_phy_param`  in `bt_conn_le_phy_update()` (`CONFIG_BT_USER_PHY_UPDATE=y`) with preferred coding scheme or default to S=8 coding in connections when not using `bt_conn_le_phy_update()` (and to stay in Coded PHY, `CONFIG_BT_AUTO_PHY_UPDATE=n` is required by applications).

Also, restricting the `CONFIG_BT_EXT_ADV_CODING_SELECTION` feature to "required" in Host, restricts the Host implementation use with all Controller implementations, as the BT Spec states:
```
If the Host specifies that it requires a specific coding (i.e., value 0x03 or 0x04) in the
Primary_Advertising_PHY_Options or Secondary_Advertising_PHY_Options and the
Controller supports the LE Feature (Advertising Coding Selection) but is currently
unable to provide all the required settings, then the Controller shall return the error
code Command Disallowed (0x0C).
```

To improve scheduling decisions in the Controller depending on other active roles and coexistence requirement, we should let the Host prefer a coding scheme, and Controller can decide to use S=2 instead of S=8 to reduce radio-on time.

```
When using the LE Coded PHY, it is recommended to carefully consider the impact
of radio-on time for power consumption and duty cycle for scheduling and coexistence
over the air.
```

rugeGerritsen replied:
@cvinayak , could you please elaborate on why you believe the "require" only makes sense for non-connectable advertising mentioned above? 

>To improve scheduling decisions in the Controller depending on other active roles and coexistence requirement, we should let the Host prefer a coding scheme, and Controller can decide to use S=2 instead of S=8 to reduce radio-on time.

This is already the case when no preference is provided.

cvinayak replied:
> @cvinayak , could you please elaborate on why you believe the "require" only makes sense for non-connectable advertising mentioned above?

As stated already here https://github.com/zephyrproject-rtos/zephyr/pull/85234#discussion_r1944814494 for a connectable advertising, a connection established can only attain a "preferred" coding scheme. Requiring a coding scheme for connectable advertising does not make sense, as Controller defaults to S=8 or to a "preferred" S=2 using user initiated PHY update.

>This is already the case when no preference is provided.

By specification, "no preference" will default to S=8 by the Controller, Host will not be able to prefer a S=2. Requiring S=2 should work with a Controller as it is better for scheduling, but spec does state that a Controller return error when unable to provide the settings (I can not think of today why requiring S2 could be rejected by a Controller, but I am just going by the Spec.)

rugeGerritsen replied:
>As stated already here https://github.com/zephyrproject-rtos/zephyr/pull/85234#discussion_r1944814494 for a connectable advertising, a connection established can only attain a "preferred" coding scheme. Requiring a coding scheme for connectable advertising does not make sense, as Controller defaults to S=8 or to a "preferred" S=2 using user initiated PHY update.

This feature is about what coding scheme used for advertising packets only - what is used on the ACL is independent.

>By specification, "no preference" will default to S=8 by the Controller, 

The spec states "The Host has no preferred or required coding when transmitting on the LE Coded PHY". This allows the controller to choose either S2 or S8

cvinayak replied:
> This feature is about what coding scheme used for advertising packets only - what is used on the ACL is independent.

Yes, that is why I said in the earlier comments that ""require" PHY makes sense only for non-connectable advertising.".
I cannot think of a usecase where applications "require" S=2 (shorter range) for connectable advertising, makes sense for non-connectable usecases like mesh to reduce latency.

> This allows the controller to choose either S2 or S8

Spec states S8 "shall"  be assumed:
```
If advertising on the LE Coded PHY, the S=8 coding shall be assumed
unless the current advertising parameters require the use of S=2 for an advertising
physical channel, in which case the S=2 coding shall be assumed for that advertising
physical channel.
```

I have no objections to not have an option for Host to prefer a coding scheme, but please update the Kconfig help text accordingly.

rugeGerritsen replied:
> Spec states S8 "shall" be assumed:

The extract from the spec pasted there is about advertising data length time validation. It doesn't state that the controller shall use S8

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -780,6 +784,14 @@ config BT_CTLR_ADV_PERIODIC_RSP
+	default y if BT_EXT_ADV_CODING_SELECTION
+	help
+	  Enable support for Bluetooth 6.0 Advertising Coding Selection
+	  in the controller.
```
```suggestion
	  in the Controller.
```

Thomas-Deppe committed:
Bluetooth: Host: Add support for Advertising Coding Selection

Adds API for Advertising Coding Selection.

Introduces two new advertising options to configure the advertiser's
requirement concerning coding scheme when LE Coded PHY is configured.
While the Bluetooth v6.0 specification makes a distinction betweeen
preferred and required advertising PHY options, a simplification is
made to only expose the required PHY options. Inline with how LE Coded
PHY is implemented; this API will set both the primary and secondary
advertising PHY's to the same coding scheme.

The support is enabled by CONFIG_BT_EXT_ADV_CODING_SELECTION, and requires
a controller that selects CONFIG_BT_CTLR_ADV_EXT_CODING_SELECTION_SUPPORT.

Signed-off-by: Thomas Deppe <thomas.deppe@nordicsemi.no>

Thomas-Deppe commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1171,6 +1191,22 @@ static int le_ext_adv_param_set(struct bt_le_ext_adv *adv,
 		cp->sec_adv_phy = BT_HCI_LE_PHY_CODED;
+
+#if defined(CONFIG_BT_EXT_ADV_CODING_SELECTION)
+		if (opcode == BT_HCI_OP_LE_SET_EXT_ADV_PARAM_V2) {
```
I have done this now :) Thank you!

jhedberg commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1121,20 +1121,40 @@ static int le_ext_adv_param_set(struct bt_le_ext_adv *adv,
+	struct bt_hci_cp_le_set_ext_adv_param_v2 *cp;
+#else
 	struct bt_hci_cp_le_set_ext_adv_param *cp;
+#endif /*CONFIG_BT_EXT_ADV_CODING_SELECTION*/
```
I think it's ok to keep it in `hci_types.h` since it's part of the HCI spec.

Thomas-Deppe commented:
Sorry for the update after the approvals. We have decided to leave out the prefers option and just keep requirement as a simplification. Push was to address this [comment](https://github.com/zephyrproject-rtos/zephyr/pull/85234#discussion_r1944825321). Thanks again for all the feedback :)
